packages/firestore/src/local/lru_garbage_collector.ts

```diff
@@ -0,0 +1,268 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { QueryData } from './query_data';
+import { PersistenceTransaction } from './persistence';
+import { PersistencePromise } from './persistence_promise';
+import { ListenSequenceNumber } from '../core/types';
+import { ListenSequence } from '../core/listen_sequence';
+import { assert } from '../util/assert';
+import { AnyJs } from '../util/misc';
+
+/**
+ * Persistence layers intending to use LRU Garbage collection should have reference delegates that
+ * implement this interface. This interface defines the operations that the LRU garbage collector
+ * needs from the persistence layer.
+ */
+export interface LruDelegate {
+  getTargetCount(txn: PersistenceTransaction): PersistencePromise<number>;
+
+  /** Enumerates all the targets in the QueryCache. */
+  forEachTarget(
+    txn: PersistenceTransaction,
+    f: (target: QueryData) => void
+  ): PersistencePromise<void>;
+
+  /** Enumerates sequence numbers for documents not associated with a target. */
+  forEachOrphanedDocumentSequenceNumber(
+    txn: PersistenceTransaction,
+    f: (sequenceNumber: ListenSequenceNumber) => void
+  ): PersistencePromise<void>;
+
+  /**
+   * Removes all targets that have a sequence number less than or equal to `upperBound`, and are not
+   * present in the `activeTargetIds` set.
+   *
+   * @return the number of targets removed.
+   */
+  removeTargets(
+    txn: PersistenceTransaction,
+    upperBound: ListenSequenceNumber,
+    activeTargetIds: LiveTargets
+  ): PersistencePromise<number>;
+
+  /**
+   * Removes all unreferenced documents from the cache that have a sequence number less than or
+   * equal to the given `upperBound`.
+   *
+   * @return the number of documents removed.
+   */
+  removeOrphanedDocuments(
+    txn: PersistenceTransaction,
+    upperBound: ListenSequenceNumber
+  ): PersistencePromise<number>;
+}
+
+/**
+ * Describes an object whose keys are active target ids. We do not care about the type of the values.
+ */
+export type LiveTargets = {
+  [id: number]: AnyJs;
+};
+
+/**
+ * A selective port of `java.util.PriorityQueue`
+ * {@see <a href="https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/PriorityQueue.java">PriorityQueue.java</a>}
+ * The queue does not grow and must have an initial capacity when it is constructed. Additionally, elements may only be
+ * `poll()`'d and cannot be removed from any other position in the queue.
+ */
+class PriorityQueue<T> {
```

Comment:
Yes, I need to store duplicates.

I'm on the fence about this. Using `SortedSet` definitely *feels* like a hack (and would have more allocations), but if we're worried about code size, I think it could work as you describe. I'll see if I can get it working, then I guess we can decide between the two.